### PR TITLE
Separate templates from contentpage

### PR DESCRIPTION
**Pull request by fritzbrand**
Separating Templates out from ContentPages, so they can be managed as standalone entities

### CODE REVIEW

File: home/admin.py [hunk 1]
```diff
@@ -6,3 +6,8 @@
 @admin.register(PageView)
 class PageViewAdmin(admin.ModelAdmin):
     list_display = [field.name for field in PageView._meta.get_fields()]
+
+
+# @admin.register(WhatsAppTemplate)
+# class WhatsappTemplateAdmin(admin.ModelAdmin):
+#     list_display = [field.name for field in WhatsAppTemplate._meta.get_fields()]
```
Comment (jerith) [inline home/admin.py:13]: Do we need this? If not, let's delete it. (If we're not sure yet, maybe add a TODO to clean it up so we don't forget.)

File: home/api.py [hunk 2]
```diff
@@ -174,12 +179,46 @@ def get_queryset(self):
...
+    known_query_parameters = BaseAPIViewSet.known_query_parameters.union(["page", "qa"])
+    pagination_class = PageNumberPagination
+    search_fields = ["name", "body"]
+    filter_backends = (SearchFilter,)
+
+    def get_queryset(self):
+        qa = self.request.query_params.get("qa")
+
+        if qa:
+            # return the latest revision for each OrderedContentSet
```
Comment (jerith) [inline home/api.py:198]: OrderedContentSet?

File: home/migrations/0066_alter_whatsapptemplate_category.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
+# Generated by Django 4.2.11 on 2024-03-22 09:02
```
Comment (jerith) [inline home/migrations/0066_alter_whatsapptemplate_category.py:1]: Can we squash all these migrations into one before we merge? (And maybe do it occasionally along the way to avoid unnecessary bloat?)

File: home/models.py [hunk 4]
```diff
@@ -944,7 +945,7 @@ def clean(self):
 
 
 @receiver(pre_save, sender=ContentPage)
-def update_embedding(sender, instance, *args, **kwargs):
+def update_template_embedding(sender, instance, *args, **kwargs):
```
Comment (jerith) [inline home/models.py:948]: We've renamed the `ContentPage` version to `update_template_embedding` and added a `WhatsAppTemplate` version called `update_embedding`. Can we change the names to something less confusing? Do we even need this for WhatsAppTemplate at all?

File: home/models.py [hunk 5]
```diff
@@ -1215,3 +1216,285 @@ class PageView(models.Model):
...
+    #     else:
+    #         # do_stuff_on_publish()
+    #         print("THIS IS A PUBLISH")
+    #     return super().save(*args, **kwargs)
+
+    def __str__(self):
+        """String repr of this snippet."""
+        return self.name
+
+    class Meta:  # noqa
```
Comment (jerith) [inline home/models.py:1339]: Should `Meta` be closer to the top of the class?

File: home/models.py [hunk 6]
```diff
@@ -1215,3 +1216,285 @@ class PageView(models.Model):
...
+                )
+            )
+
+        if errors:
+            raise ValidationError(errors)
+
+        return result
+
+    @property
+    def whatsapp_template_fields(self):
```
Comment (jerith) [inline home/models.py:1434]: This appears to serve that same purpose as the `fields` property above. Do we need both? If so, what's the difference?

File: home/wagtail_hooks.py [hunk 7]
```diff
@@ -104,7 +106,7 @@ class ContentPageAdmin(ModelAdmin):
         "related_pages",
         "parental",
     )
-    list_filter = ("locale",)
+
```
Comment (jerith) [inline home/wagtail_hooks.py:109]: Is this change intentional?

File: home/wagtail_hooks.py [hunk 8]
```diff
@@ -115,7 +117,7 @@ class ContentPageAdmin(ModelAdmin):
         "viber_body",
         "slug",
     )
-    list_export = ("locale", "title")
+    list_export = "title"
```
Comment (jerith) [inline home/wagtail_hooks.py:120]: Is this change intentional?

File: home/whatsapp.py [hunk 9]
```diff
@@ -112,44 +112,102 @@ def create_whatsapp_template(
...
-        "Content-Type": "application/json",
-    }
+    if image_id is not None:
+        image_obj = Image.objects.get(id=image_id)
+        components.append(create_whatsapp_template_image(image_obj))
+
+    submit_whatsapp_template(name, category, locale, components)
+
+
+# TODO: Get correct type for below
```
Comment (jerith) [inline home/whatsapp.py:126]: Obsolete TODO.

File: home/admin.py (hunk 1 above)
Comment (fritzbrand) [reply home/admin.py:13]: Not needed, removed

File: home/api.py (hunk 2 above)
Comment (fritzbrand) [reply home/api.py:198]: Fixed

File: home/migrations/0066_alter_whatsapptemplate_category.py (hunk 3 above)
Comment (fritzbrand) [reply home/migrations/0066_alter_whatsapptemplate_category.py:1]: Cool, squashed.  Believe it or not, I have squashed it a few times already but forgot to do it before this push.  Thx

File: home/models.py (hunk 4 above)
Comment (fritzbrand) [reply home/models.py:948]: Cool, for now, I renamed them to make more sense.  I don't know enough of it to make a call whether we need it or not, but I know we have a ticket in play on removing it, so will leave it for that one

File: home/models.py [hunk 10]
```diff
@@ -1215,3 +1216,285 @@ class PageView(models.Model):
...
+    category = models.CharField(
+        max_length=14,
+        choices=Category.choices,
+        default=Category.MARKETING,
+    )
+    quick_replies = ClusterTaggableManager(
+        through="home.TemplateQuickReplyContent", blank=True
+    )
+
+    related_pages = StreamField(
```
[the remark this answers is not in the document]
Comment (fritzbrand) [reply home/models.py:1259]: Have added a TODO to check with Rudi on this, but the way I see it, we need it there in order to migrate things over, but I am not completely convinced of that approach.  Good catch, will resolve with Rudi

File: home/models.py [hunk 11]
```diff
@@ -1215,3 +1216,285 @@ class PageView(models.Model):
...
+        max_length=4096,
+    )
+
+    example_values = StreamField(
+        [
+            (
+                "example_values",
+                blocks.CharBlock(
+                    label="Example Value",
+                ),
```
[the remark this answers is not in the document]
Comment (fritzbrand) [reply home/models.py:1289]: I tried, but black reformatted it like this.  Anything else I can try?

File: home/models.py (hunk 5 above)
Comment (fritzbrand) [reply home/models.py:1339]: Yup, moved

File: home/models.py [hunk 12]
```diff
@@ -1215,3 +1216,285 @@ class PageView(models.Model):
...
+        return revision
+
+    def clean(self):
+        result = super().clean()
+        errors = {}
+
+        # The WA title is needed for all templates to generate a name for the template
+        if not self.name:
+            errors.setdefault("name", []).append(
+                ValidationError("All WhatsApp templates need a title.")
```
[the remark this answers is not in the document]
Comment (fritzbrand) [reply home/models.py:1384]: Pages used titles, I changed it to name as it makes more sense.  Updated

File: home/models.py [hunk 13]
```diff
@@ -1215,3 +1216,285 @@ class PageView(models.Model):
...
+            )
+
+        message = self.message
+
+        right_mismatch = re.findall(r"(?<!\{){[^{}]*}\}", message)
+        left_mismatch = re.findall(r"\{{[^{}]*}(?!\})", message)
```
[the remark this answers is not in the document]
Comment (fritzbrand) [reply home/models.py:1390]: I don't know on this one,  It was there on like 922 for ContentPages, so I just copied it for Standalone templates. It has to do with checking that we have the same amount of opening and closing braces when we work with placeholders, but I don't understand it deep enough to suggest another method right now.  Any ideas?

File: home/models.py [hunk 14]
```diff
@@ -1215,3 +1216,285 @@ class PageView(models.Model):
...
+                    f"Please provide numeric variables only. You provided {non_digit_variables}."
+                )
+            )
+
+        # Check variable order
+        actual_digit_variables = [var for var in vars_in_msg if var.isdecimal()]
+        expected_variables = [str(j + 1) for j in range(len(actual_digit_variables))]
+        if actual_digit_variables != expected_variables:
+            errors.setdefault("message", []).append(
+                StreamBlockValidationError(
```
[the remark this answers is not in the document]
Comment (fritzbrand) [reply home/models.py:1415]: I removed the nesting.  I think Mudiwa did these validations, will check with him.  Added a TODO for now

File: home/wagtail_hooks.py (hunk 7 above)
Comment (fritzbrand) [reply home/wagtail_hooks.py:109]: Nope, I put it back

File: home/wagtail_hooks.py (hunk 8 above)
Comment (fritzbrand) [reply home/wagtail_hooks.py:120]: Nope, fixed

File: home/models.py (hunk 6 above)
Comment (fritzbrand) [reply home/models.py:1427]: Nah, I've remove `fields` this replaces it

File: home/serializers.py [hunk 15]
```diff
@@ -460,3 +460,8 @@ class OrderedContentSetSerializer(BaseSerializer):
     name = NameField(read_only=True)
     pages = OrderedPagesField(read_only=True)
     profile_fields = ProfileFieldsField(read_only=True)
+
+
+class WhatsAppTemplateSerializer(BaseSerializer):
+    name = NameField(read_only=True)
+    body = BodyField(read_only=True)
```
[the remark this answers is not in the document]
Comment (fritzbrand) [reply home/serializers.py:467]: Nah it's not, was thinking I can figure this out as part of the import/export ticket, but will need help either way

File: home/models.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (jerith) [reply home/models.py:1262]: Oh, right. It currently lives on the message. I think the correct place for it is on the "template message" or whatever we end up with for using templates in pages.

File: home/models.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (jerith) [reply home/models.py:1292]: Black treats a trailing comma as an instruction to keep something multi-line. Try remove those and rerun?

File: home/models.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (jerith) [reply home/models.py:1379]: Let's add a TODO to either replace or explain and worry about it later.

File: home/models.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (jerith) [reply home/models.py]: The message is buried in a stream thingy in the ContentPage version, so that one needs the nesting. This one doesn't.

File: home/models.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (fritzbrand) [reply home/models.py:1262]: Removed related pages field

File: home/api.py [hunk 16]
```diff
@@ -174,12 +179,46 @@ def get_queryset(self):
...
 
 
+class WhatsAppTemplateViewset(BaseAPIViewSet):
+    model = WhatsAppTemplate
+    base_serializer_class = WhatsAppTemplateSerializer
+    listing_default_fields = BaseAPIViewSet.listing_default_fields + [
+        "name",
+        "body",
+    ]
+    known_query_parameters = BaseAPIViewSet.known_query_parameters.union(["page", "qa"])
```
Comment (rudigiesler) [inline home/api.py:189]: I don't think `page` is a query param we use here, or at least it's not being handled in `get_queryset`

File: home/models.py [hunk 17]
```diff
@@ -1268,3 +1269,251 @@ class PageView(models.Model):
     )
     message = models.IntegerField(blank=True, default=None, null=True)
     data = models.JSONField(default=dict, blank=True, null=True)
+
+
+class TemplateContentQuickReply(TagBase):
+
+    class Meta:
+        verbose_name = "quick reply2"
+        verbose_name_plural = "quick replie2"
```
Comment (rudigiesler) [inline home/models.py:1278]: Is there a reason for this naming, I don't think that `verbose_name` is used as an identifier anywhere, it's just what's shown on the front end

File: home/models.py [hunk 18]
```diff
@@ -1268,3 +1269,251 @@ class PageView(models.Model):
...
+        Only submits if the create templates is enabled
+        and if the template fields are different to the previous revision
+        """
+        if not settings.WHATSAPP_CREATE_TEMPLATES:
+            return
+
+        # If there are any missing fields in the previous revision, then carry on
+        if previous_revision:
+            previous_revision = previous_revision.as_object()
+            previous_revision_fields = previous_revision.whatsapp_template_fields
```
Comment (rudigiesler) [inline home/models.py:1481]: I don't see any `whatsapp_template_fields` on the model, just `fields`, so maybe we want that instead?

File: home/serializers.py [hunk 19]
```diff
@@ -460,3 +460,9 @@ class OrderedContentSetSerializer(BaseSerializer):
     name = NameField(read_only=True)
     pages = OrderedPagesField(read_only=True)
     profile_fields = ProfileFieldsField(read_only=True)
+
+
+# TODO: Figure out what fields to make part of this serialiser
+class WhatsAppTemplateSerializer(BaseSerializer):
+    name = NameField(read_only=True)
+    body = BodyField(read_only=True)
```
Comment (rudigiesler) [inline home/serializers.py:468]: I don't think that `body` is a field on the template, and we probably want all fields from the model here (unless they're being included in the page API instead)

File: home/wagtail_hooks.py [hunk 20]
```diff
@@ -219,6 +222,89 @@ class OrderedContentSetViewSet(SnippetViewSet):
...
+#                 # FieldPanel("image"),
+#                 # FieldPanel("message"),
+#                 # FieldPanel("quick_replies", heading="Quick Replies"),
+#                 # FieldPanel("locale"),
+#                 # FieldPanel("example_values"),
+#             ],
+#             heading="Template Example Values",
+#         ),
+#     ]
+# register_snippet(ExampleValuesViewSet)
```
Comment (rudigiesler) [inline home/wagtail_hooks.py:307]: We can probably clean up all this commented out code in this file

**Review by rudigiesler — COMMENTED** [review]
A few comments, luckily all minor, well done :)

File: home/serializers.py (hunk 19 above)
Comment (fritzbrand) [reply home/serializers.py:468]: Yea that's what the TODO is for, we will update the serializer once we have this new "submit via button" thing implemented, or that's what jerith and I discussed anyway

File: home/models.py (hunk 18 above)
Comment (fritzbrand) [reply home/models.py:1481]: Yes we have renamed it back to `fields`

File: home/models.py (hunk 17 above)
Comment (fritzbrand) [reply home/models.py]: Nah this was me testing something and forgot to change it back.  Updated

File: home/api.py (hunk 16 above)
Comment (fritzbrand) [reply home/api.py]: Cool, removed it

File: home/serializers.py (hunk 19 above)
Comment (fritzbrand) [reply home/serializers.py]: Actually, doesn't look like we actually need the serializer yet, which is why nothing broke with it in that state.  Removed it for now, will add back once it actually gets used

File: home/models.py (hunk 18 above)
Comment (rudigiesler) [reply home/models.py:1481]: Then I think this is only passing tests because we're never calling this function with a previous revision (it's always `submit_whatsapp_template(None)`), because I think this needs to be `previous_revision_fields = previous_revision.fields`

File: home/models.py [hunk 21]
```diff
@@ -1268,3 +1269,252 @@ class PageView(models.Model):
...
+        else:
+            previous_revision_fields = ()
+
+        if self.fields == previous_revision_fields:
+            return
+
+        self.template_name = self.create_whatsapp_template_name()
+        print("MODELS")
+        print(self.example_values)
```
Comment (rudigiesler) [inline home/models.py:1490]: I think we can remove these prints

File: home/tests/test_models.py [hunk 22]
```diff
@@ -642,3 +647,173 @@ def test_clean_text_char_limit(self):
...
+
+        request = responses.calls[0].request
+        assert json.loads(request.body) == {
+            "category": "UTILITY",
+            "components": [{"text": "Test WhatsApp Message 1", "type": "BODY"}],
+            "language": "en_US",
+            "name": f"wa_title_{wat.get_latest_revision().id}",
+        }
+
+    # TODO: translate all these
```
Comment (rudigiesler) [inline home/tests/test_models.py:743]: I think this is done

File: home/tests/test_models.py [hunk 23]
```diff
@@ -642,3 +647,173 @@ def test_clean_text_char_limit(self):
...
+
+    # TODO: Find a better way to test quick replies
+    @override_settings(WHATSAPP_CREATE_TEMPLATES=True)
+    @responses.activate
+    def test_template_create_with_buttons(self):
+        # page = create_page(is_whatsapp_template=True, has_quick_replies=True)
```
Comment (rudigiesler) [inline home/tests/test_models.py:749]: We can remove this

**Review by rudigiesler — COMMENTED** [review]
Just some things that need to be cleaned up

File: home/tests/test_models.py (hunk 23 above)
Comment (fritzbrand) [reply home/tests/test_models.py:749]: Urgh, sorry, Done

File: home/tests/test_models.py (hunk 22 above)
Comment (fritzbrand) [reply home/tests/test_models.py:743]: Yup, removed

File: home/models.py (hunk 21 above)
Comment (fritzbrand) [reply home/models.py]: Done, soz

**Review by rudigiesler — APPROVED** [review]

File: home/models.py (hunk 18 above)
Comment (jerith) [reply home/models.py:1481]: When we make template submission its own action instead of something that happens when we save a revision, none of this comparison will need to happen anymore. I think it's fine to leave it broken for now and remove it entirely in the next PR.

Comment (fritzbrand) [reply home/models.py:1481]: Sorry I missed this comment from Rudi after mine saying we renamed it back to `fields`

**Review by rudigiesler — APPROVED** [review]
